lib/src/git_subprocess.rs

```diff
@@ -0,0 +1,484 @@
+    GitExternalError(String),
+    #[error("Failed to convert path {0:?} to a string")]
+    PathConversionError(std::path::PathBuf),
+}
+impl From<std::io::Error> for GitError {
+    fn from(value: std::io::Error) -> Self {
+        GitError::GitForkError(IoError(value))
+    }
+}
+
+impl From<GitError> for GitFetchError {
+    fn from(value: GitError) -> Self {
+        match value {
+            GitError::NoSuchRemote(remote) => GitFetchError::NoSuchRemote(remote),
+            GitError::GitNotFound(git_path) => GitFetchError::GitNotFound(git_path),
+            GitError::GitForkError(io_err) => GitFetchError::GitForkError(io_err.0),
+            GitError::GitExternalError(error) => GitFetchError::GitExternalError(error),
+            GitError::PathConversionError(path) => GitFetchError::PathConversionError(path),
+        }
+    }
+}
+
+impl From<GitError> for GitPushError {
+    fn from(value: GitError) -> Self {
+        match value {
+            GitError::NoSuchRemote(remote) => GitPushError::NoSuchRemote(remote),
+            GitError::GitNotFound(git_path) => GitPushError::GitNotFound(git_path),
+            GitError::GitForkError(io_err) => GitPushError::GitForkError(io_err),
+            GitError::GitExternalError(error) => GitPushError::GitExternalError(error),
+            GitError::PathConversionError(path) => GitPushError::PathConversionError(path),
+        }
+    }
+}
+
+/// Context for creating git subprocesses
+pub(crate) struct GitSubprocessContext<'a> {
+    git_dir: String,
+    work_tree: String,
+    git_path: &'a str,
+}
+
+impl<'a> GitSubprocessContext<'a> {
+    pub(crate) fn new(git_dir: impl ToString, work_tree: impl ToString, git_path: &'a str) -> Self {
+        GitSubprocessContext {
+            git_dir: git_dir.to_string(),
+            work_tree: work_tree.to_string(),
+            git_path,
+        }
+    }
+
+    pub(crate) fn from_git2(
+        git_repo: &git2::Repository,
+        git_path: &'a str,
+    ) -> Result<Self, GitError> {
+        let git_dir = git_repo
+            .path()
+            .to_str()
+            .ok_or_else(|| GitError::PathConversionError(git_repo.path().to_path_buf()))?;
+        let work_tree = work_tree_from_git_dir(git_repo.path())
+            .map(|p| {
+                p.to_str()
+                    .map(|x| x.to_string())
+                    .ok_or(GitError::PathConversionError(p))
+            })
+            .map_err(GitError::GitExternalError)??;
+
+        Ok(Self::new(git_dir, work_tree, git_path))
+    }
+
+    fn spawn(&self, args: &[String], stdout: Stdio) -> Result<Output, GitError> {
+        let full_args = {
+            let mut v = Vec::with_capacity(args.len() + 2);
+            v.push(format!("--git-dir={}", self.git_dir));
+            v.push(format!("--work-tree={}", self.work_tree));
+            v.extend_from_slice(args);
+            v
+        };
+
+        tracing::debug!("shelling out to `{} {}`", self.git_path, args.join(" "));
```

Comment:
nit: we can also prints the `Command` struct, which contains all information we need for debug trace.

```rust
let mut cmd = Command::new(..);
...
tracing::debug!(?cmd, ...);
cmd.spawn()
```